synapseformation/client.py

```diff
@@ -0,0 +1,87 @@
+"""Synapse Formation client"""
+import synapseclient
+from synapseclient import Synapse
+
+from .create import SynapseCreation
+from . import utils
+
+
+# def expand_config(config: dict) -> dict:
+#     """Expands shortened configuration to the official json format"""
+#     # TODO: Once all resources are either under a key mapping or lists
+#     # this will have to chance
+#     if isinstance(config, dict) and config.get('type') == "Project":
+#         # TODO: figure out how to clean this up.
+#         # Remove yaml anchor objects as they are expanded
+#         to_remove = [key for key in config
+#                      if key not in ["type", "name", "children"]]
+#         for key in to_remove:
+#             config.pop(key)
+#         # Get children if exists
+#         children = config.get('children')
+#         if children is not None:
+#             expand_config(children)
+#     else:
+#         # Loop through folders and create them
+#         for index, folder in enumerate(config):
+#             # The folder configuration can be lists or dicts
+#             # Must pull out children if it exists
+#             if isinstance(folder, dict):
+#                 children = folder.get('children')
+#             else:
+#                 config[index] = {"name": folder,
+#                                  "type": "Folder"}
+#                 children = None
+
+#             # Create nested folders
+#             if children is not None:
+#                 expand_config(children)
+#     return config
+
+
+def _create_synapse_resources(syn: Synapse, config: dict,
+                              parentid: str = None):
+    """Recursively steps through template and creates synapse resources
+
+    Args:
+        syn: Synapse connection
+        config: Synapse Formation template dict
+        parentid: Synapse folder or project id to store entities
+    """
+    creation_cls = SynapseCreation(syn)
+    if isinstance(config, dict) and config.get('type') == "Project":
+        project = creation_cls.get_or_create_project(name=config['name'])
```

Comment:
What if the template provides both a name and a Synapse ID for a project? Should we preferentially use the ID over the name since it's considered more stable even they can both uniquely identify a project? 

I wonder if the `SynapseCreation` class should have a more generic `get_or_create()` method that takes `**config` as input and it figures it out what to create from the mandatory `type` field. This would help simplify the logic in this function (I think). Anyways, just a thought.

Reply:
That is a good idea.  I definitely think we could have a `determine_synapse_resource` function or something that determines what function that needs to be called.  That being said, this sort of logic has to live here or in that function.  I think we can open a separate ticket to improve the underlying class.